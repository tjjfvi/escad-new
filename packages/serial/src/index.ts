/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-non-null-assertion */

import ieee754 from "ieee754"
import { createHash, Hash as Hasher } from "crypto"

type Hash = string

export interface SerializeOptions {
  chunkSize?: number,
  chunkMinSize?: number,
  hashMap?: WeakMap<object, Hash>,
}

enum Kind {
  null = 0,
  end = 1,
  string = 2,
  number = 3,
  false = 4,
  true = 5,
  object = 6,
  array = 7,
  buffer = 8,
  undefined = 9,
}

const endMarker = Symbol()

export async function* serialize(
  rootValue: unknown,
  {
    chunkSize = 1 << 14, // 16kb
    chunkMinSize = chunkSize / 2,
    hashMap,
  }: SerializeOptions = {},
){
  const valueMemo = new Map<unknown, number>()
  const hashMemo = new Map<Hash, number>()
  const idToHash = new Map<number, Hash>()
  let idN = 1
  let currentChunk: Buffer | undefined
  let currentInd = 0
  let totalPosition = 0
  const hasherStack: [value: object, hasher: Hasher, id: number, start: number][] = []
  const stack = [rootValue]
  while(stack.length) {
    const start = totalPosition
    const value = stack.pop()
    if(value === endMarker && hashMap) {
      const [value, hasher, id, start] = hasherStack.pop()!
      const hash = hasher.digest("hex")
      hasherStack[hasherStack.length - 1]?.[1].update(hash, "hex")
      const hashMemoId = hashMemo.get(hash)
      hashMap.set(value, hash)
      if(hashMemoId && unwrite(totalPosition - start)) {
        idN = id
        yield* writeId(hashMemoId)
        continue
      }
      else {
        hashMemo.set(hash, id)
        idToHash.set(id, hash)
      }
    }
    const memoId = valueMemo.get(value)
    if(memoId !== undefined) {
      yield* writeId(memoId)
      if(hashMap) {
        const hash = idToHash.get(memoId)
        if(hash) hasherStack[hasherStack.length - 1]?.[1].update(hash, "hex")
      }
      continue
    }
    if(hashMap && typeof value === "object" && value) {
      const hash = hashMap.get(value)
      const hashMemoId = hashMemo.get(hash!)
      if(hash && hashMemoId) {
        yield* writeId(hashMemoId)
        hasherStack[hasherStack.length - 1]?.[1].update(hash, "hex")
        continue
      }
    }
    const id = idN++
    let hasher = hashMap && createHash("sha256")
    let deferHasher = false
    yield* writeId(0)
    valueMemo.set(value, id)
    if(typeof value === "object" || typeof value === "function")
      if(Array.isArray(value)) {
        yield* writeKind(Kind.array, hasher)
        stack.push(endMarker, ...value)
        deferHasher = true
      }
      else if(value instanceof Buffer) {
        yield* writeKind(Kind.buffer, hasher)
        yield* write(4, buf => buf.writeUInt32LE(value.length), hasher)
        yield* writeBuffer(value, hasher)
      }
      else if(value === null)
        yield* writeKind(Kind.null, hasher)
      else {
        yield* writeKind(Kind.object, hasher)
        stack.push(endMarker)
        for(const key in value)
          stack.push(value[key as never], key)
        deferHasher = true
      }
    else if(typeof value === "string") {
      yield* writeKind(Kind.string, hasher)
      yield* write(4, buf => buf.writeUInt32LE(value.length), hasher)
      yield* write(value.length, buf => buf.write(value, "utf8"), hasher)
    }
    else if(typeof value === "number") {
      yield* writeKind(Kind.number, hasher)
      yield* write(8, buf => ieee754.write(buf, value, 0, true, 52, 8), hasher)
    }
    else if(value === undefined)
      yield* writeKind(Kind.undefined, hasher)
    else if(typeof value === "boolean")
      yield* writeKind(value ? Kind.true : Kind.false, hasher)
    else if(value === endMarker)
      yield* writeKind(Kind.end)
    else
      throw new Error(`Cannot serialize value of type "${typeof value}"`)

    if(hasher && !deferHasher) {
      const hash = hasher.digest("hex")
      idToHash.set(id, hash)
      hasherStack[hasherStack.length - 1]?.[1].update(hash, "hex")
    }

    if(hasher && deferHasher)
      hasherStack.push([value as object, hasher, id, start])
  }

  if(currentChunk)
    yield currentChunk.slice(0, currentInd)

  return

  function unwrite(length: number){
    if(length > currentInd)
      return false
    currentInd -= length
    totalPosition -= length
    return true
  }

  function writeKind(kind: Kind, hasher?: Hasher){
    return write(1, buf => buf.writeUInt8(kind), hasher)
  }

  function writeId(id: number){
    return write(4, buf => buf.writeUInt32LE(id))
  }

  async function* writeBuffer(buffer: Buffer, hasher?: Hasher){
    hasher?.update(buffer)

    if(buffer.length > chunkMinSize) {
      if(currentChunk)
        yield currentChunk.slice(0, currentInd)
      yield buffer
      currentChunk = undefined
      return
    }

    yield* write(buffer.length, buf => buffer.copy(buf))
  }

  async function* write(length: number, cb: (buffer: Buffer) => void, hasher?: Hasher){
    if(length > chunkSize) {
      if(currentChunk)
        yield currentChunk.slice(0, currentInd)
      const buffer = Buffer.allocUnsafe(length)
      cb(buffer)
      hasher?.update(buffer)
      yield buffer
      currentChunk = undefined
      return
    }

    if(currentChunk && currentInd + length > currentChunk.length) {
      yield currentChunk.slice(0, currentInd)
      currentChunk = undefined
    }

    if(!currentChunk) {
      currentChunk = Buffer.allocUnsafe(chunkSize)
      currentInd = 0
    }

    const section = currentChunk.slice(currentInd, currentInd + length)
    cb(section)
    hasher?.update(section)

    currentInd += length
    totalPosition += length
  }
}

export interface DeserializeOptions {

}

export async function deserialize(
  stream: AsyncIterable<Buffer>,
  {}: DeserializeOptions = {},
){
  type Target = [unknown[]] | [Record<string, unknown>, string | undefined]
  const memo = new Map<number, unknown>()
  let idN = 1
  const iterator = stream[Symbol.asyncIterator]()
  const readQueue: Buffer[] = []
  let readQueueLength = 0
  const targetStack: Target[] = [[[]]]
  do {
    const target = targetStack[targetStack.length - 1]
    const value = await readValue()
    if(value === endMarker) {
      targetStack.pop()
      continue
    }
    if(target.length === 1)
      target[0].push(value)
    else if(target[1] !== undefined) {
      target[0][target[1]] = value
      target[1] = undefined
    }
    else {
      if(typeof value !== "string")
        throw new Error("Invalid key for object")
      target[1] = value
    }
  } while(targetStack.length > 1)

  return (targetStack[0][0] as unknown[])[0]

  async function readValue(): Promise<unknown>{
    let id = await readId()
    if(id)
      return memo.get(id)
    id = idN++
    const value = await _readValue()
    memo.set(id, value)
    return value
  }

  async function _readValue(): Promise<unknown>{
    const kind = await readKind()
    if(kind === Kind.array) {
      const value: unknown[] = []
      targetStack.push([value])
      return value
    }
    if(kind === Kind.object) {
      const value = {}
      targetStack.push([value, undefined])
      return value
    }
    if(kind === Kind.end)
      return endMarker
    if(kind === Kind.number)
      return ieee754.read(await read(8), 0, true, 52, 8)
    if(kind === Kind.null)
      return null
    if(kind === Kind.undefined)
      return undefined
    if(kind === Kind.string)
      return (await read((await read(4)).readUInt32LE())).toString("utf8")
    if(kind === Kind.true)
      return true
    if(kind === Kind.false)
      return false
    if(kind === Kind.buffer)
      return await read((await read(4)).readUInt32LE())
    throw new Error(`Invalid kind ${kind}`)
  }

  async function readKind(): Promise<Kind>{
    const kind = await read(1).then(buf => buf[0])
    return kind
  }

  async function readId(){
    return await read(4).then(buf => buf.readUInt32LE())
  }

  async function read(length: number){
    while(length > readQueueLength) {
      const result = await iterator.next()
      if(result.done) throw new Error("Unexpected EOF")
      readQueue.push(result.value)
      readQueueLength += result.value.length
    }

    readQueueLength -= length
    if(length === readQueue[0].length)
      return readQueue.shift()!
    if(length < readQueue[0].length) {
      const buffer = readQueue[0].slice(0, length)
      readQueue[0] = readQueue[0].slice(length)
      return buffer
    }
    const buffer = Buffer.allocUnsafe(length)
    let writtenLength = 0
    while(writtenLength < length) {
      if(length - writtenLength >= readQueue[0].length) {
        writtenLength += readQueue[0].length
          readQueue.shift()!.copy(buffer, writtenLength)
          continue
      }
      readQueue[0].copy(buffer, writtenLength, 0, length - writtenLength)
      readQueue[0] = readQueue[0].slice(writtenLength - length)
      break
    }
    return buffer
  }
}
