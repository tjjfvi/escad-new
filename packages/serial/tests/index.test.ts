
import { serialize, deserialize } from "../src"

let objects: unknown[] = [
  0,
  null,
  Math.SQRT1_2,
  NaN,
  Number.MAX_SAFE_INTEGER,
  Number.MAX_VALUE,
  Number.MIN_SAFE_INTEGER,
  Number.MIN_VALUE,
  +Infinity,
  -Infinity,
  undefined,
  "hello world",
  JSON.stringify({ hello: "world" }),
  // eslint-disable-next-line max-len
  "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent in sem porttitor, egestas lacus eu, aliquet mi. Maecenas laoreet, purus in malesuada finibus, odio lacus finibus felis, et iaculis quam dolor vel velit. Morbi vel ornare purus. Phasellus et lacus tempor, fringilla lectus nec, ullamcorper justo. Duis id porttitor massa, vel facilisis ligula. Vestibulum vitae pretium sem. Curabitur diam diam, finibus quis efficitur ut, sodales eget mi. Vestibulum sed egestas lectus. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae",
  [...Array(128)].map((_, i) => String.fromCharCode(i)),
  [...Array(128)].map((_, i) => String.fromCharCode(i)).join(""),
  Buffer.from([...Array(256).keys()]),
]

objects.push({
  this: {
    is: {
      a: {
        long: {
          chain: {
            of: {
              properties: {
                that: {
                  is: {
                    not: {
                      circular: NaN,
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  },
})

const hashMap0 = new Map()

const hashes0 = objects.map(x => serialize.hash(x, new Map()))
const hashes1 = objects.map(x => serialize.hash(x, hashMap0))
const hashes2 = objects.map(x => serialize.hash(x, hashMap0))

test("hashes", () => {
  expect(hashes0).toMatchSnapshot()
  expect(hashes1).toStrictEqual(hashes0)
  expect(hashes2).toStrictEqual(hashes1)
})

const hashMap1 = new Map()
const serialized = [...serialize(objects, { hashMap: hashMap1 })]

test("serialized", () =>
  expect(
    "\n" + serialized
      .map(x => x.toString("hex"))
      .join("")
      .replace(/.{64}/g, "$&\n"),
  ).toMatchSnapshot(),
)

test("hashMap1", () =>
  expect(hashMap1).toMatchSnapshot(),
)

const hashMap2 = new Map()
const deserialized = deserialize.withHash((async function*(){
  for(const value of serialized)
    yield value
})(), { hashMap: hashMap2 })

test("deserialized", async () =>
  expect((await deserialized).value).toStrictEqual(objects),
)

test("hashMap2", async () => {
  await deserialized
  expect(hashMap2).toMatchSnapshot()
})

test("deserializeHash", async () => {
  await deserialized
  expect((await deserialized).hash).toEqual(hashMap1.get(objects))
})
