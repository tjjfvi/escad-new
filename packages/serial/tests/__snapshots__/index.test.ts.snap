// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`errors on circular 1`] = `"Attempted to serialize circular value"`;

exports[`hashMap1 1`] = `
Map {
  Object {
    "hello": "world",
  } => "b788d18f764f15baab21684b3a97e7fe65be8612866d51916962430d271b17d8",
  Object {
    "hello": "world",
  } => "b788d18f764f15baab21684b3a97e7fe65be8612866d51916962430d271b17d8",
  Object {
    "circular": NaN,
  } => "829583d46ceba929dc1c2107930337ee06df17c860da609722dbfd4f6d41cc73",
  Object {
    "not": Object {
      "circular": NaN,
    },
  } => "84cad4337b1695cd31bf706800f0b98498e230af9ae4f204a98e9c6f4e5d5ef9",
  Object {
    "is": Object {
      "not": Object {
        "circular": NaN,
      },
    },
  } => "3f427c3ea440f9738b928a27e66930e1d2a8c5dd36c78ccd9349b80aef6ab23f",
  Object {
    "that": Object {
      "is": Object {
        "not": Object {
          "circular": NaN,
        },
      },
    },
  } => "18b3dc8bf0411f6eb017563f725126b06a3ec2f9252a2904358d790dff100853",
  Object {
    "properties": Object {
      "that": Object {
        "is": Object {
          "not": Object {
            "circular": NaN,
          },
        },
      },
    },
  } => "27fb3f5cec84cca5e4c030bbca9f97e00a32282a1a4316b687033ee04bfaf9ff",
  Object {
    "of": Object {
      "properties": Object {
        "that": Object {
          "is": Object {
            "not": Object {
              "circular": NaN,
            },
          },
        },
      },
    },
  } => "07e3e5be3ed5407b4f088518ad1b167fed8a946aeb0e2d0ae772980345709d34",
  Object {
    "chain": Object {
      "of": Object {
        "properties": Object {
          "that": Object {
            "is": Object {
              "not": Object {
                "circular": NaN,
              },
            },
          },
        },
      },
    },
  } => "549c1b8ec3b994a1ef93d47322e62585478b78680179d258925420fe7c73496d",
  Object {
    "long": Object {
      "chain": Object {
        "of": Object {
          "properties": Object {
            "that": Object {
              "is": Object {
                "not": Object {
                  "circular": NaN,
                },
              },
            },
          },
        },
      },
    },
  } => "97ba50e944520fd5b0034c5dd3b319175bba4a3bb1f9afed88670332c32c7b60",
  Object {
    "a": Object {
      "long": Object {
        "chain": Object {
          "of": Object {
            "properties": Object {
              "that": Object {
                "is": Object {
                  "not": Object {
                    "circular": NaN,
                  },
                },
              },
            },
          },
        },
      },
    },
  } => "b072a44ccb9bb7eb0b07b2ac364f3c3b6d2eadc6bcd51ae3bc6c96382c457661",
  Object {
    "is": Object {
      "a": Object {
        "long": Object {
          "chain": Object {
            "of": Object {
              "properties": Object {
                "that": Object {
                  "is": Object {
                    "not": Object {
                      "circular": NaN,
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  } => "9fdfdd95fbbbb932a37733c109d278335045789416c4ae6e7d2d25f7971008f0",
  Object {
    "this": Object {
      "is": Object {
        "a": Object {
          "long": Object {
            "chain": Object {
              "of": Object {
                "properties": Object {
                  "that": Object {
                    "is": Object {
                      "not": Object {
                        "circular": NaN,
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  } => "c16db9a6b4a19fabeb3d3ab0fa512587db0d6e7972baee2b8ae610ae2cd5ad51",
  Array [
    0,
    null,
    0.7071067811865476,
    NaN,
    9007199254740991,
    1.7976931348623157e+308,
    -9007199254740991,
    5e-324,
    Infinity,
    -Infinity,
    undefined,
    "hello world",
    "{\\"hello\\":\\"world\\"}",
    Object {
      "hello": "world",
    },
    Object {
      "hello": "world",
    },
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent in sem porttitor, egestas lacus eu, aliquet mi. Maecenas laoreet, purus in malesuada finibus, odio lacus finibus felis, et iaculis quam dolor vel velit. Morbi vel ornare purus. Phasellus et lacus tempor, fringilla lectus nec, ullamcorper justo. Duis id porttitor massa, vel facilisis ligula. Vestibulum vitae pretium sem. Curabitur diam diam, finibus quis efficitur ut, sodales eget mi. Vestibulum sed egestas lectus. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae",
    Object {
      "data": Array [
        0,
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10,
        11,
        12,
        13,
        14,
        15,
        16,
        17,
        18,
        19,
        20,
        21,
        22,
        23,
        24,
        25,
        26,
        27,
        28,
        29,
        30,
        31,
        32,
        33,
        34,
        35,
        36,
        37,
        38,
        39,
        40,
        41,
        42,
        43,
        44,
        45,
        46,
        47,
        48,
        49,
        50,
        51,
        52,
        53,
        54,
        55,
        56,
        57,
        58,
        59,
        60,
        61,
        62,
        63,
        64,
        65,
        66,
        67,
        68,
        69,
        70,
        71,
        72,
        73,
        74,
        75,
        76,
        77,
        78,
        79,
        80,
        81,
        82,
        83,
        84,
        85,
        86,
        87,
        88,
        89,
        90,
        91,
        92,
        93,
        94,
        95,
        96,
        97,
        98,
        99,
        100,
        101,
        102,
        103,
        104,
        105,
        106,
        107,
        108,
        109,
        110,
        111,
        112,
        113,
        114,
        115,
        116,
        117,
        118,
        119,
        120,
        121,
        122,
        123,
        124,
        125,
        126,
        127,
        128,
        129,
        130,
        131,
        132,
        133,
        134,
        135,
        136,
        137,
        138,
        139,
        140,
        141,
        142,
        143,
        144,
        145,
        146,
        147,
        148,
        149,
        150,
        151,
        152,
        153,
        154,
        155,
        156,
        157,
        158,
        159,
        160,
        161,
        162,
        163,
        164,
        165,
        166,
        167,
        168,
        169,
        170,
        171,
        172,
        173,
        174,
        175,
        176,
        177,
        178,
        179,
        180,
        181,
        182,
        183,
        184,
        185,
        186,
        187,
        188,
        189,
        190,
        191,
        192,
        193,
        194,
        195,
        196,
        197,
        198,
        199,
        200,
        201,
        202,
        203,
        204,
        205,
        206,
        207,
        208,
        209,
        210,
        211,
        212,
        213,
        214,
        215,
        216,
        217,
        218,
        219,
        220,
        221,
        222,
        223,
        224,
        225,
        226,
        227,
        228,
        229,
        230,
        231,
        232,
        233,
        234,
        235,
        236,
        237,
        238,
        239,
        240,
        241,
        242,
        243,
        244,
        245,
        246,
        247,
        248,
        249,
        250,
        251,
        252,
        253,
        254,
        255,
      ],
      "type": "Buffer",
    },
    Object {
      "this": Object {
        "is": Object {
          "a": Object {
            "long": Object {
              "chain": Object {
                "of": Object {
                  "properties": Object {
                    "that": Object {
                      "is": Object {
                        "not": Object {
                          "circular": NaN,
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  ] => "37b92107d19584a4882d2f6a1432bfcb01ebbfcc25a49dff5b956f814220197d",
}
`;

exports[`hashMap2 1`] = `
Map {
  Object {
    "hello": "world",
  } => "b788d18f764f15baab21684b3a97e7fe65be8612866d51916962430d271b17d8",
  Object {
    "data": Array [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      15,
      16,
      17,
      18,
      19,
      20,
      21,
      22,
      23,
      24,
      25,
      26,
      27,
      28,
      29,
      30,
      31,
      32,
      33,
      34,
      35,
      36,
      37,
      38,
      39,
      40,
      41,
      42,
      43,
      44,
      45,
      46,
      47,
      48,
      49,
      50,
      51,
      52,
      53,
      54,
      55,
      56,
      57,
      58,
      59,
      60,
      61,
      62,
      63,
      64,
      65,
      66,
      67,
      68,
      69,
      70,
      71,
      72,
      73,
      74,
      75,
      76,
      77,
      78,
      79,
      80,
      81,
      82,
      83,
      84,
      85,
      86,
      87,
      88,
      89,
      90,
      91,
      92,
      93,
      94,
      95,
      96,
      97,
      98,
      99,
      100,
      101,
      102,
      103,
      104,
      105,
      106,
      107,
      108,
      109,
      110,
      111,
      112,
      113,
      114,
      115,
      116,
      117,
      118,
      119,
      120,
      121,
      122,
      123,
      124,
      125,
      126,
      127,
      128,
      129,
      130,
      131,
      132,
      133,
      134,
      135,
      136,
      137,
      138,
      139,
      140,
      141,
      142,
      143,
      144,
      145,
      146,
      147,
      148,
      149,
      150,
      151,
      152,
      153,
      154,
      155,
      156,
      157,
      158,
      159,
      160,
      161,
      162,
      163,
      164,
      165,
      166,
      167,
      168,
      169,
      170,
      171,
      172,
      173,
      174,
      175,
      176,
      177,
      178,
      179,
      180,
      181,
      182,
      183,
      184,
      185,
      186,
      187,
      188,
      189,
      190,
      191,
      192,
      193,
      194,
      195,
      196,
      197,
      198,
      199,
      200,
      201,
      202,
      203,
      204,
      205,
      206,
      207,
      208,
      209,
      210,
      211,
      212,
      213,
      214,
      215,
      216,
      217,
      218,
      219,
      220,
      221,
      222,
      223,
      224,
      225,
      226,
      227,
      228,
      229,
      230,
      231,
      232,
      233,
      234,
      235,
      236,
      237,
      238,
      239,
      240,
      241,
      242,
      243,
      244,
      245,
      246,
      247,
      248,
      249,
      250,
      251,
      252,
      253,
      254,
      255,
    ],
    "type": "Buffer",
  } => "542621a6352d9fe2961a61df73613a004bfe44010b04119fdc71278967ee9d10",
  Object {
    "circular": NaN,
  } => "829583d46ceba929dc1c2107930337ee06df17c860da609722dbfd4f6d41cc73",
  Object {
    "not": Object {
      "circular": NaN,
    },
  } => "84cad4337b1695cd31bf706800f0b98498e230af9ae4f204a98e9c6f4e5d5ef9",
  Object {
    "is": Object {
      "not": Object {
        "circular": NaN,
      },
    },
  } => "3f427c3ea440f9738b928a27e66930e1d2a8c5dd36c78ccd9349b80aef6ab23f",
  Object {
    "that": Object {
      "is": Object {
        "not": Object {
          "circular": NaN,
        },
      },
    },
  } => "18b3dc8bf0411f6eb017563f725126b06a3ec2f9252a2904358d790dff100853",
  Object {
    "properties": Object {
      "that": Object {
        "is": Object {
          "not": Object {
            "circular": NaN,
          },
        },
      },
    },
  } => "27fb3f5cec84cca5e4c030bbca9f97e00a32282a1a4316b687033ee04bfaf9ff",
  Object {
    "of": Object {
      "properties": Object {
        "that": Object {
          "is": Object {
            "not": Object {
              "circular": NaN,
            },
          },
        },
      },
    },
  } => "07e3e5be3ed5407b4f088518ad1b167fed8a946aeb0e2d0ae772980345709d34",
  Object {
    "chain": Object {
      "of": Object {
        "properties": Object {
          "that": Object {
            "is": Object {
              "not": Object {
                "circular": NaN,
              },
            },
          },
        },
      },
    },
  } => "549c1b8ec3b994a1ef93d47322e62585478b78680179d258925420fe7c73496d",
  Object {
    "long": Object {
      "chain": Object {
        "of": Object {
          "properties": Object {
            "that": Object {
              "is": Object {
                "not": Object {
                  "circular": NaN,
                },
              },
            },
          },
        },
      },
    },
  } => "97ba50e944520fd5b0034c5dd3b319175bba4a3bb1f9afed88670332c32c7b60",
  Object {
    "a": Object {
      "long": Object {
        "chain": Object {
          "of": Object {
            "properties": Object {
              "that": Object {
                "is": Object {
                  "not": Object {
                    "circular": NaN,
                  },
                },
              },
            },
          },
        },
      },
    },
  } => "b072a44ccb9bb7eb0b07b2ac364f3c3b6d2eadc6bcd51ae3bc6c96382c457661",
  Object {
    "is": Object {
      "a": Object {
        "long": Object {
          "chain": Object {
            "of": Object {
              "properties": Object {
                "that": Object {
                  "is": Object {
                    "not": Object {
                      "circular": NaN,
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  } => "9fdfdd95fbbbb932a37733c109d278335045789416c4ae6e7d2d25f7971008f0",
  Object {
    "this": Object {
      "is": Object {
        "a": Object {
          "long": Object {
            "chain": Object {
              "of": Object {
                "properties": Object {
                  "that": Object {
                    "is": Object {
                      "not": Object {
                        "circular": NaN,
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  } => "c16db9a6b4a19fabeb3d3ab0fa512587db0d6e7972baee2b8ae610ae2cd5ad51",
  Array [
    0,
    null,
    0.7071067811865476,
    NaN,
    9007199254740991,
    1.7976931348623157e+308,
    -9007199254740991,
    5e-324,
    Infinity,
    -Infinity,
    undefined,
    "hello world",
    "{\\"hello\\":\\"world\\"}",
    Object {
      "hello": "world",
    },
    Object {
      "hello": "world",
    },
    "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent in sem porttitor, egestas lacus eu, aliquet mi. Maecenas laoreet, purus in malesuada finibus, odio lacus finibus felis, et iaculis quam dolor vel velit. Morbi vel ornare purus. Phasellus et lacus tempor, fringilla lectus nec, ullamcorper justo. Duis id porttitor massa, vel facilisis ligula. Vestibulum vitae pretium sem. Curabitur diam diam, finibus quis efficitur ut, sodales eget mi. Vestibulum sed egestas lectus. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae",
    Object {
      "data": Array [
        0,
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10,
        11,
        12,
        13,
        14,
        15,
        16,
        17,
        18,
        19,
        20,
        21,
        22,
        23,
        24,
        25,
        26,
        27,
        28,
        29,
        30,
        31,
        32,
        33,
        34,
        35,
        36,
        37,
        38,
        39,
        40,
        41,
        42,
        43,
        44,
        45,
        46,
        47,
        48,
        49,
        50,
        51,
        52,
        53,
        54,
        55,
        56,
        57,
        58,
        59,
        60,
        61,
        62,
        63,
        64,
        65,
        66,
        67,
        68,
        69,
        70,
        71,
        72,
        73,
        74,
        75,
        76,
        77,
        78,
        79,
        80,
        81,
        82,
        83,
        84,
        85,
        86,
        87,
        88,
        89,
        90,
        91,
        92,
        93,
        94,
        95,
        96,
        97,
        98,
        99,
        100,
        101,
        102,
        103,
        104,
        105,
        106,
        107,
        108,
        109,
        110,
        111,
        112,
        113,
        114,
        115,
        116,
        117,
        118,
        119,
        120,
        121,
        122,
        123,
        124,
        125,
        126,
        127,
        128,
        129,
        130,
        131,
        132,
        133,
        134,
        135,
        136,
        137,
        138,
        139,
        140,
        141,
        142,
        143,
        144,
        145,
        146,
        147,
        148,
        149,
        150,
        151,
        152,
        153,
        154,
        155,
        156,
        157,
        158,
        159,
        160,
        161,
        162,
        163,
        164,
        165,
        166,
        167,
        168,
        169,
        170,
        171,
        172,
        173,
        174,
        175,
        176,
        177,
        178,
        179,
        180,
        181,
        182,
        183,
        184,
        185,
        186,
        187,
        188,
        189,
        190,
        191,
        192,
        193,
        194,
        195,
        196,
        197,
        198,
        199,
        200,
        201,
        202,
        203,
        204,
        205,
        206,
        207,
        208,
        209,
        210,
        211,
        212,
        213,
        214,
        215,
        216,
        217,
        218,
        219,
        220,
        221,
        222,
        223,
        224,
        225,
        226,
        227,
        228,
        229,
        230,
        231,
        232,
        233,
        234,
        235,
        236,
        237,
        238,
        239,
        240,
        241,
        242,
        243,
        244,
        245,
        246,
        247,
        248,
        249,
        250,
        251,
        252,
        253,
        254,
        255,
      ],
      "type": "Buffer",
    },
    Object {
      "this": Object {
        "is": Object {
          "a": Object {
            "long": Object {
              "chain": Object {
                "of": Object {
                  "properties": Object {
                    "that": Object {
                      "is": Object {
                        "not": Object {
                          "circular": NaN,
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  ] => "37b92107d19584a4882d2f6a1432bfcb01ebbfcc25a49dff5b956f814220197d",
}
`;

exports[`hashes 1`] = `
Array [
  "82f63b935a385c7a6e1176aec3307db0df131cd469b491b666e01d47d91a02d1",
  "df3f619804a92fdb4057192dc43dd748ea778adc52bc498ce80524c014b81119",
  "012027978e0e1ca800ce8942d01ea90190616c81437e8a42edee6401e902ca19",
  "04bbc158e897f9b2c81288c02f1356ea45229750233af42e36553ca37b4c1dc0",
  "5ad2bc1c1351c969f49f8df02e7dbe2791cd74df007a2ac49397b48515d2ee60",
  "07544043976d5d4d7b0365bd998d8c74441bf00cb448905ff0d535c4a8b162ab",
  "78898bfb9ffd6d56b7aff5e81f59b553ba8c36fa7c79b00a23f65530b4fca289",
  "02e6d00eac41da596a79bbd5f8bc152e83ce4e1c79065682ac32464d192deefc",
  "427cffbf16297d7316283231812649c641304dfd365e3bd0e2337d8940046886",
  "e93544448cc3cd6b579650146d9e67ee3c6f2b196973d9f956ad7cf9285ed69a",
  "9f076b7eb7fdc0311cd3208cdbbebbf8014dd3a05e35191c96947b358a362b40",
  "02a0a69aec177e29eb4dc1249b88b808f3f998b58397b4344b57358e0d0c13c8",
  "af420092584128d3ffc4b08441f0e21e04c358aa355fa143a640c0b7f941d24e",
  "b788d18f764f15baab21684b3a97e7fe65be8612866d51916962430d271b17d8",
  "b788d18f764f15baab21684b3a97e7fe65be8612866d51916962430d271b17d8",
  "2a9e5021bf3d7bf2f4c069ab5327406e8dad5c4fda23e06991b128c36c7942d7",
  "542621a6352d9fe2961a61df73613a004bfe44010b04119fdc71278967ee9d10",
  "c16db9a6b4a19fabeb3d3ab0fa512587db0d6e7972baee2b8ae610ae2cd5ad51",
]
`;

exports[`serialized 1`] = `
"
070000000300000000000000000000000000000003000000cd3b7f669ea0e63f
03000000010000000000f07f03000000ffffffffffff3f4303000000ffffffff
ffffef7f03000000ffffffffffff3fc303000000010000000000000003000000
000000000000f07f03000000000000000000f0ff09000000020000000b000000
68656c6c6f20776f726c6402000000110000007b2268656c6c6f223a22776f72
6c64227d06000000020000000500000068656c6c6f0200000005000000776f72
6c64010000002f000000020000003a0200004c6f72656d20697073756d20646f
6c6f722073697420616d65742c20636f6e736563746574757220616469706973
63696e6720656c69742e205072616573656e7420696e2073656d20706f727474
69746f722c2065676573746173206c616375732065752c20616c697175657420
6d692e204d616563656e6173206c616f726565742c20707572757320696e206d
616c6573756164612066696e696275732c206f64696f206c616375732066696e
696275732066656c69732c20657420696163756c6973207175616d20646f6c6f
722076656c2076656c69742e204d6f7262692076656c206f726e617265207075
7275732e2050686173656c6c7573206574206c616375732074656d706f722c20
6672696e67696c6c61206c6563747573206e65632c20756c6c616d636f727065
72206a7573746f2e204475697320696420706f72747469746f72206d61737361
2c2076656c20666163696c69736973206c6967756c612e20566573746962756c
756d207669746165207072657469756d2073656d2e2043757261626974757220
6469616d206469616d2c2066696e696275732071756973206566666963697475
722075742c20736f64616c65732065676574206d692e20566573746962756c75
6d207365642065676573746173206c65637475732e20566573746962756c756d
20616e746520697073756d207072696d697320696e206661756369627573206f
726369206c756374757320657420756c74726963657320706f73756572652063
7562696c69612063757261650800000000010000000102030405060708090a0b
0c0d0e0f101112131415161718191a1b1c1d1e1f202122232425262728292a2b
2c2d2e2f303132333435363738393a3b3c3d3e3f404142434445464748494a4b
4c4d4e4f505152535455565758595a5b5c5d5e5f606162636465666768696a6b
6c6d6e6f707172737475767778797a7b7c7d7e7f808182838485868788898a8b
8c8d8e8f909192939495969798999a9b9c9d9e9fa0a1a2a3a4a5a6a7a8a9aaab
acadaeafb0b1b2b3b4b5b6b7b8b9babbbcbdbebfc0c1c2c3c4c5c6c7c8c9cacb
cccdcecfd0d1d2d3d4d5d6d7d8d9dadbdcdddedfe0e1e2e3e4e5e6e7e8e9eaeb
ecedeeeff0f1f2f3f4f5f6f7f8f9fafbfcfdfeff060000000200000004000000
7468697306000000020000000200000069730600000002000000010000006106
00000002000000040000006c6f6e67060000000200000005000000636861696e
0600000002000000020000006f6606000000020000000a00000070726f706572
7469657306000000020000000400000074686174060000003800000006000000
02000000030000006e6f7406000000020000000800000063697263756c617225
0000003200000032000000320000003200000032000000320000003200000032
00000032000000320000003200000032000000"
`;
